### ISSUE TITLE
IOException after updating from 3.0.0 to 3.2.0

### ISSUE THREAD

**Issue by aryono01** [wdullaer/MaterialDateTimePicker#349]
Hello, I encounter this exception after updating from 3.0.0 to 3.2.0. Some of my app user experienced it and I haven't able to reproduce it myself. (log taken from fabric)

```
Fatal Exception: java.lang.RuntimeException: Parcelable encountered IOException writing serializable object (name = com.wdullaer.materialdatetimepicker.date.DefaultDateRangeLimiter)
       at android.os.Parcel.writeSerializable(Parcel.java:1394)
       at android.os.Parcel.writeValue(Parcel.java:1341)
       at android.os.Parcel.writeArrayMapInternal(Parcel.java:644)
       at android.os.BaseBundle.writeToParcelInner(BaseBundle.java:1313)
       at android.os.Bundle.writeToParcel(Bundle.java:1034)
       at android.os.Parcel.writeBundle(Parcel.java:669)
       at android.app.FragmentState.writeToParcel(Fragment.java:144)
       at android.os.Parcel.writeTypedArray(Parcel.java:1197)
       at android.app.FragmentManagerState.writeToParcel(FragmentManager.java:381)
       at android.os.Parcel.writeParcelable(Parcel.java:1363)
       at android.os.Parcel.writeValue(Parcel.java:1268)
       at android.os.Parcel.writeArrayMapInternal(Parcel.java:644)
       at android.os.BaseBundle.writeToParcelInner(BaseBundle.java:1313)
       at android.os.Bundle.writeToParcel(Bundle.java:1034)
       at android.os.Parcel.writeBundle(Parcel.java:669)
       at android.app.ActivityManagerProxy.activityStopped(ActivityManagerNative.java:3232)
       at android.app.ActivityThread$StopInfo.run(ActivityThread.java:3653)
       at android.os.Handler.handleCallback(Handler.java:815)
       at android.os.Handler.dispatchMessage(Handler.java:104)
       at android.os.Looper.loop(Looper.java:194)
       at android.app.ActivityThread.main(ActivityThread.java:5763)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:960)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
Caused by java.io.NotSerializableException: com.wdullaer.materialdatetimepicker.date.DatePickerDialog
       at java.io.ObjectOutputStream.writeNewObject(ObjectOutputStream.java:1344)
       at java.io.ObjectOutputStream.writeObjectInternal(ObjectOutputStream.java:1651)
       at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1497)
       at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1461)
       at java.io.ObjectOutputStream.writeFieldValues(ObjectOutputStream.java:959)
       at java.io.ObjectOutputStream.defaultWriteObject(ObjectOutputStream.java:360)
       at java.io.ObjectOutputStream.writeHierarchy(ObjectOutputStream.java:1054)
       at java.io.ObjectOutputStream.writeNewObject(ObjectOutputStream.java:1384)
       at java.io.ObjectOutputStream.writeObjectInternal(ObjectOutputStream.java:1651)
       at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1497)
       at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1461)
       at android.os.Parcel.writeSerializable(Parcel.java:1389)
       at android.os.Parcel.writeValue(Parcel.java:1341)
       at android.os.Parcel.writeArrayMapInternal(Parcel.java:644)
       at android.os.BaseBundle.writeToParcelInner(BaseBundle.java:1313)
       at android.os.Bundle.writeToParcel(Bundle.java:1034)
       at android.os.Parcel.writeBundle(Parcel.java:669)
       at android.app.FragmentState.writeToParcel(Fragment.java:144)
       at android.os.Parcel.writeTypedArray(Parcel.java:1197)
       at android.app.FragmentManagerState.writeToParcel(FragmentManager.java:381)
       at android.os.Parcel.writeParcelable(Parcel.java:1363)
       at android.os.Parcel.writeValue(Parcel.java:1268)
       at android.os.Parcel.writeArrayMapInternal(Parcel.java:644)
       at android.os.BaseBundle.writeToParcelInner(BaseBundle.java:1313)
       at android.os.Bundle.writeToParcel(Bundle.java:1034)
       at android.os.Parcel.writeBundle(Parcel.java:669)
       at android.app.ActivityManagerProxy.activityStopped(ActivityManagerNative.java:3232)
       at android.app.ActivityThread$StopInfo.run(ActivityThread.java:3653)
       at android.os.Handler.handleCallback(Handler.java:815)
       at android.os.Handler.dispatchMessage(Handler.java:104)
       at android.os.Looper.loop(Looper.java:194)
       at android.app.ActivityThread.main(ActivityThread.java:5763)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:960)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
```

Here's my code:
```
final int year = CALENDAR.get(Calendar.YEAR);
                final int month = CALENDAR.get(Calendar.MONTH);
                final int date = CALENDAR.get(Calendar.DATE);

                final DatePickerDialog dialog = DatePickerDialog.newInstance(new DatePickerDialog.OnDateSetListener() {
                    @Override
                    public void onDateSet(DatePickerDialog view, int i, int i1, int i2) {
                        CALENDAR.set(i, i1, i2);
                        tanggal.setText(Helper.getReadableDate(CALENDAR));
                    }
                }, year, month, date);

                dialog.setMinDate(Helper.getTomorrowDate());
                dialog.show(getBaseActivity().getFragmentManager(), "datepicker");
```

**Comment by wdullaer** [comment]
I haven't been able to reproduce this.
I was planning to implement Parcelable on this class, rather than relying on serializable, which hopefully should make this go away (because it won't be relying on jvm magic that much)

If you do manage to reproduce it, please let me know (I'm guessing some special value of minDate triggers this)

**Comment by aryono01** [comment]
Thank you for quick reply. I'm still investigating this issue and will post here when i found something. Thank you for creating and sharing this great library!

**Comment by MustafaMonir** [comment]
Hello,
this happens anytime the app pauses (ex: pressing home button, or moving to another app) while the date dialog is showing.

**Comment by wdullaer** [comment]
That reproduces it, thanks.
I'll look into it!

**Comment by laalto** [comment]
Here's a workaround I applied to the codebase I was working on - basically dismiss the datepicker dialog if the activity hosting the dialogfragment is paused. There's a side effect that the dialog is not restored but that is acceptable for UX in this case.

```
        activity.getApplication().registerActivityLifecycleCallbacks(new Application.ActivityLifecycleCallbacks() {
            @Override
            public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
            }

            @Override
            public void onActivityStarted(Activity activity) {
            }

            @Override
            public void onActivityResumed(Activity activity) {
            }

            @Override
            public void onActivityPaused(Activity activity) {
                datePickerDialog.dismiss();
                activity.getApplication().unregisterActivityLifecycleCallbacks(this);
            }

            @Override
            public void onActivityStopped(Activity activity) {
            }

            @Override
            public void onActivitySaveInstanceState(Activity activity, Bundle outState) {
            }

            @Override
            public void onActivityDestroyed(Activity activity) {
            }
        });

        datePickerDialog.show(activity.getFragmentManager(), "Datepickerdialog");
```

**Comment by laalto** [comment]
Thanks, that's a simpler (less code fat) way to achieve the same workaround.

**Comment by wdullaer** [comment]
This is fixed in v3.2.1
The DefaultDateRangePicker could not be serialized leading to the Exception.
If it wouldn't have thrown the Exception this would have introduced a memory leak.

I decided to change the Interface to require Parcelable rather than Serializable. This is a bit more work if you want to implement it, but will force you to think about which values to save. It would be all too easy to try and serialize the Context in the implementation of this Interface
